linter_plugin.py

```diff
@@ -1,31 +1,52 @@
-"""Certbot ACME PyLint plugin.
-
+"""Certbot PyLint plugin.
 http://docs.pylint.org/plugins.html
-
 """
-from astroid import MANAGER
-from astroid import nodes
-
-
-def register(unused_linter):
-    """Register this module as PyLint plugin."""
-    pass
-
-
-def _transform(cls):
-    # fix the "no-member" error on instances of
-    # letsencrypt.acme.util.ImmutableMap subclasses (instance
-    # attributes are initialized dynamically based on __slots__)
-
-    # TODO: this is too broad and applies to any tested class...
-
-    #if cls.slots() is not None:
-    #    for slot in cls.slots():
-    #        cls.locals[slot.value] = [nodes.EmptyNode()]
-
-    if cls.name == 'JSONObjectWithFields':
-        # _fields is magically introduced by JSONObjectWithFieldsMeta
-        cls.locals['_fields'] = [nodes.EmptyNode()]
-
-
-MANAGER.register_transform(nodes.Class, _transform)
+from pylint.checkers import BaseChecker
+from pylint.interfaces import IAstroidChecker
+from pylint.checkers.utils import check_messages
+
+
+# Module in theses packages can import os module.
+WHITELIST_PACKAGES = ['acme', 'certbot_compatibility_test', 'letshelp_certbot', 'lock_test']
+
+
+class ForbidStandardOsModule(BaseChecker):
```

Comment:
Is this code based on checker for `deprecated-modules`? If so, can you state that in a comment here and add a link to the pylint code?

Reply:
No, this code is based on official documentation: http://pylint.pycqa.org/en/stable/how_tos/custom_checkers.html

Reply:
To help me verify this code, I went and looked at the [deprecated-module checker](https://github.com/PyCQA/pylint/blob/b20a2984c94e2946669d727dbda78735882bf50a/pylint/checkers/imports.py#L287). One thing the code you have here doesn't prevent is `import os.path` or `from os.path import foo`. If we wanted to check for this, doing so should be as simple as calling `startswith("os.")` on the string you're comparing with the constant "os" below.

I personally think we should do this. I don't think we have to as I'm not aware of any problematic functionality in the `os.path` module, however, I think it's cleaner and more Pythonic to have 1 way to do things which would be to use `certbot.compat.os.path` instead of also having the option to just use `os.path` (or telling people to not use `certbot.compat.os.path` and to make an exception for the `path` module and to use that directly instead.)

I think this means we should also fix up `certbot.compat.os` so `import certbot.compat.os.path` works like it does for the `os` module, but I think it's probably worth it to have a single path to access the code there.

What do you think?

If we do this, I think it's worth adding a comment that this class is loosely based on the checker for the deprecated-module warning in `pylint` and maybe adding a link to it. I think this gives us an implementation of very similar checks we can refer back to if we ever need to change the code here in the future.

Reply:
Ah, you are right! I forgot this situation. Of course, I totally agree, and implemented the additional control toward `os.path`-like imports.

I adapted also `certbot.compat.os` to automatically register a `certbot.compat.os.path`, to be used in a similar way than `os.path`.

Add a comment in the linter to have a similar example to study, as you said.